### PR TITLE
[confmap] fix bug of unmarshalling slice values

### LINKED ISSUE #4001: Config unmarshalling doesn't work properly for slice values
**Describe the bug**
If we have a default config for a component where there is a slice value with 2 and more item, there is no way to override that value with a smaller slice.

The problem is that [mapstructure.decodeMap func](https://github.com/mitchellh/mapstructure/blob/4664f9ee512b4dfa71b501c4960d469c205b2cf1/mapstructure.go#L1046) that we use for unmarshalling compares elements of the slice one by one and does not override the length of the result. It leads to the problem that we cannot override a default slice value with a slice of smaller size. 

This is an unexpected behaviour for users. We should be able to fully override default slice values.

**Steps to reproduce**
Real example is https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/dotnetdiagnosticsreceiver configuration:

Try to override the `counters` value with a 1-item slice like this:
```
receivers:
  dotnet_diagnostics:
    counters: ["MyCounters"]
``` 

**What did you expect to see?**
A slice config value should be fully overridden. In the example above, we should get `counters: ["MyCounters"]`

**What did you see instead?**
Only subset of the default slice items are overridden, the rest of the default slice stays the same. In the example above, we get `counters: ["MyCounters", "Microsoft.AspNetCore.Hosting"]`

### PR DESCRIPTION
**Pull request by fatsheep9146**
**Description:** 
fix the bug of unmarshalling slice values for package confmap

**Link to tracking Issue:** 
fix #4001

### CODE REVIEW

**Comment by fatsheep9146** [comment]
@dmitryax could you help review this pr?

**Comment by dmitryax** [comment]
Please address comments from @bogdandrutu in https://github.com/open-telemetry/opentelemetry-collector/pull/6163. I think we should not override with `nil` but probably should override with `[]`

**Comment by fatsheep9146** [comment]
> Please address comments from @bogdandrutu in #6163. I think we should not override with `nil` but probably should override with `[]`

@dmitryax 

I think it is already doing like that, in unit test, these two cases are also covered

https://github.com/open-telemetry/opentelemetry-collector/blob/4dc71d6716c20667412e3c2262be22b14db1ce8a/confmap/confmap_test.go#L449-L472

- The default value will not be overrided by `nil`.
- The default value will be overrided by `[]` empty slice.

File: confmap/confmap.go [hunk 1]
```diff
@@ -336,3 +337,34 @@ type Marshaler interface {
...
+//
+// 3. configuration have `keys` field and have nil value for this key, the output should be default config
+//   - for example, input is {"keys": nil}, then output is Config{ Keys: ["a", "b"]}
+//
+// 4. configuration have no `keys` field specified, the output should be default config
+//   - for example, input is {}, then output is Config{ Keys: ["a", "b"]}
+func zeroSliceHookFunc() mapstructure.DecodeHookFuncValue {
+	return func(from reflect.Value, to reflect.Value) (interface{}, error) {
+		if to.CanSet() && to.Kind() == reflect.Slice {
+			to.Set(reflect.MakeSlice(reflect.SliceOf(to.Type().Elem()), 0, 0))
```
Comment (dmitryax) [inline confmap/confmap.go:365]: Shouldn't this be?
```suggestion
			to.Set(reflect.MakeSlice(to.Type(), to.Len(), to.Cap()))
```

Comment (fatsheep9146) [reply confmap/confmap.go:365]: I think this should be `to.Set(reflect.MakeSlice(to.Type(), from.Len(), from.Cap()))`, since we use `from` to set `to`

Comment (dmitryax) [reply confmap/confmap.go]: Why `reflect.SliceOf(to.Type().Elem())`? `to.Type()` must be the slice type already

**Comment by fatsheep9146** [comment]
still doing

**Review by dmitryax — APPROVED** [review]

**Comment by fatsheep9146** [comment]
@open-telemetry/collector-maintainers 
could you help review about this pr?

**Review by codeboten — APPROVED** [review]
Thanks for including the detailed documentation @fatsheep9146

**Pull request merged**